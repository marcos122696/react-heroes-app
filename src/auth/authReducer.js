import { types } from "../types/types";

// const state = {
//     name: 'Marcos',
//     logged: true
// };

// const action = {
//     type: type.login,
//     payload: {
//         name: 'Marcos',
//         email: 'marcos@gmail.com',
//     },
// };

export const authReducer = ( state = {}, action ) => {

    switch ( action.type ) {
        case types.login:
            return {
                ...action.payload, // = name: action.payload.name
                logged: true
            }
        
        case types.logout:
            return {
                logged: false
            };
    
        default:
            return state;
    }
};