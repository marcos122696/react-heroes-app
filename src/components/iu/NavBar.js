import React, { useContext } from 'react'
import { Link, NavLink, useNavigate } from 'react-router-dom';
import { AuthContext } from '../../auth/authContext';
import { types } from '../../types/types';


export const Navbar = () => {

    const { user, dispatch } = useContext( AuthContext );
    const navigate = useNavigate();

    const handleLogout = () => {
        dispatch({
            type: types.logout
        })

        navigate('/login', {
            replace: true
        })
    };

    return (
        <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
            <Link 
                className="navbar-brand ms-3" 
                to="/"
            >
                HeroesApp
            </Link>

            <div className="navbar-collapse">
                <div className="navbar-nav">

                    <NavLink 
                        className={ ({ isActive }) => 
                            "nav-item nav-link " + ( isActive ? "active text-primary" : "" ) 
                        }
                        to="/marvel"
                    >
                        Marvel
                    </NavLink>

                    <NavLink 
                        className={ ({ isActive }) => 
                            "nav-item nav-link " + ( isActive ? "active text-primary" : "" ) 
                        }
                        to="/dc"
                    >
                        DC
                    </NavLink>

                    <NavLink 
                        className={ ({ isActive }) => 
                            "nav-item nav-link " + ( isActive ? "active text-primary" : "" ) 
                        }
                        to="/search"
                    >
                        Serch
                    </NavLink>
                </div>
            </div>

            <div className="navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end">
                <ul className="navbar-nav ml-auto">
                    
                    <span className="nav-item nav-link text-info">
                        { user.name }
                    </span>

                    <button 
                        className="nav-item nav-link btn btn-outline-secondary me-3" 
                        onClick={ handleLogout }
                    >
                        Logout
                    </button>
                </ul>
            </div>
        </nav>
    )
};