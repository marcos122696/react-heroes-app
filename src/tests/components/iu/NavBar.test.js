import { mount } from "enzyme";
import { AuthContext } from "../../../auth/authContext";
import { Navbar } from "../../../components/iu/NavBar";
import { MemoryRouter } from "react-router-dom";
import { types } from "../../../types/types";

const mockNavigate = jest.fn();
const dispatch = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}))


describe('Pruebas en el <NavBar />', () => {

    const contextValue = {
        user: {
            logged: true,
            name: 'Pedro'
        },
        dispatch: dispatch
    };

    const wrapper = mount(
        <AuthContext.Provider value={ contextValue } >
            <MemoryRouter initialEntries={ ['/navbar']}>
                <Navbar />
            </MemoryRouter>
        </AuthContext.Provider>
    );


    test('Debe de mostrarse correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.text-info').text().trim() ).toBe('Pedro');
    });


    test('Debe de llamar el logout, llamar el navigate y el dispatch con los argumentos', () => {

        wrapper.find('button').simulate('click');
        expect( mockNavigate ).toHaveBeenCalled();
        expect( mockNavigate ).toHaveBeenCalledWith('/login', {
            replace: true
        });
        expect( contextValue.dispatch ).toHaveBeenCalledWith({
            type: types.logout
        });
    });
});