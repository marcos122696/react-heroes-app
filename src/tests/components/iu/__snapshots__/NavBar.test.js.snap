// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <NavBar /> Debe de mostrarse correctamente 1`] = `
<nav
  className="navbar navbar-expand-sm navbar-dark bg-dark"
>
  <a
    className="navbar-brand ms-3"
    href="/"
    onClick={[Function]}
  >
    HeroesApp
  </a>
  <div
    className="navbar-collapse"
  >
    <div
      className="navbar-nav"
    >
      <a
        className="nav-item nav-link "
        href="/marvel"
        onClick={[Function]}
      >
        Marvel
      </a>
      <a
        className="nav-item nav-link "
        href="/dc"
        onClick={[Function]}
      >
        DC
      </a>
      <a
        className="nav-item nav-link "
        href="/search"
        onClick={[Function]}
      >
        Serch
      </a>
    </div>
  </div>
  <div
    className="navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end"
  >
    <ul
      className="navbar-nav ml-auto"
    >
      <span
        className="nav-item nav-link text-info"
      >
        Pedro
      </span>
      <button
        className="nav-item nav-link btn btn-outline-secondary me-3"
        onClick={[Function]}
      >
        Logout
      </button>
    </ul>
  </div>
</nav>
`;
